warn on future reshape alias mutation violations

dagitses opened the pull request:
warn on future reshape alias mutation violations

Summary:
Copy-on-Write Storage
=====================

Motivation
----------
PyTorch inherited from NumPy an optimization on the reshape function
that produces a view as an output if it can be represented as
such. This complicates the work of the PyTorch compilation stack
because it needs to understand the representation of the input in
order to understand if the output will be a copy or an alias.

The compilation stack would rather not be concerned with such details,
motivating the Stride Agnostic PyTorch project.

To address reshape specifically, we wish to simplify the
implementation to *always* copy, but copy lazily upon modification if
we can represent the output as a view.

Implementation plan
-------------------
We have not implemented copy-on-write tensors yet, because this is a
backward incompatible change (see Backward Incompatible section
below). But this is the design.

A copy-on-write tensor, also known as a lazily-copied tensor, is
initially going to be created by an operation like reshape which would
historically have created a view. We wish to maintain the performance
of the view but drop the aliasing aspect. Note that there is desire
for copy-on-write tensors outside of reshape, so we will likely also
want to add a public operator that can do this. It could be named
"lazy_clone" or something similar.

The core tenet of the design is that we wish to maintain the invariant
that Tensors alias if and only if they share a storage. Thus when we
create a lazy-copy, we will need to create a new storage. We also will
have to modify the source storage, since it also must now be lazily
copied.

The algorithm for creating a new copy-on-write tensor (and also
converting the source to copy-on-write) is as follows:

```
def lazy_clone(src: Tensor) -> Tensor:
    # First ensure that the source has a copy-on-write enabled storage.
    # We implement this using a custom context on the DataPtr. The
    # source tensor might already have this enabled, but if it doesn't,
    # convert it.
    if not has_copy_on_write_context(src.storage().storage_impl()):
        if not(wrap_with_copy_on_write_context(src.storage().storage_impl())):
            # For whatever reason, we weren't able to wrap the DataPtr.
            # We have to do an eager copy.
            return src.clone()

    new_storage = fork_copy_on_write_storage(src.storage()) # can't fail
    # Now just create a new tensor using the new storage.
    return new Tensor(storage=new_storage, sizes_and_strides_like=src)
```

That's the high level algorithm. The copy-on-write context that we
introduce is morally just a refcount on the underlying physical data
pointer. Each unique storage represents a set of tensors that share a
view and thus will hold a single refcount on the context.

Now we just need to intercept writes to the storage and materialize
them. We can use a few mechanisms to do this:

1) autograd knows which operators write to which tensor inputs. We can
   materialize at that point when autograd is enabled.
2) if autograd is not enabled, we can introduce a new dispatch key
   that does the same trick
3) we can also materialize whenever there's mutable access to the data
   pointer through any of `at::Tensor`, `c10::TensorImpl`,
   `c10::Storage`, `c10::StorageImpl`, or `c10::DataPtr`. With the
   current codebase, this will be too aggressive, but we will refactor
   to have a minimal set of mutable accesses.

Backwards incompatibiility
--------------------------
Changing reshape to produce a copy is a backwards incompatible change,
because users could be relying on the aliasing behavior, intentionally
or not.

For one release, rather than returning copy-on-write tensors, we
instead warn when users have triggered behavior in their program that
relies on the aliasing of the output and input.

To do this, we must simulate the behavior in a backward compatible
way. To remain backward compatible, the aliases must preserve the
invariant that they have the same storage. This is a big deviation
from the design detailed above. To get closer to the real design and
implementation, we introduce a new `c10::TensorImpl` level concept
called "Shadow Storage". The shadow storage represents what the
storage would have looked like the view actually been a lazy copy.

In the instrumented world we thus maintain the formal invariant that
tensors that alias share a storage. But we have a new invariant:
tensors that are lazy-copies of each other will share a shadow
storage.

So what do we warn on? We warn if there is a write to a tensor in a
set that shares a shadow storage followed by a read or a write to a
different tensor that shares a physical storage but has a different
shadow storage. In the real implementation, the first write would have
triggered a copy, forever cleaving the two sets of tensors, but in the
current world we instead had behavior that relied on the view-ness of
the output of reshape.

We can track these violations simply by adding a generation number to
the shadow and physical storages, updating them both on writes and
observing if a read or write ever encounters values that are out of
sync.

We have a few mechanisms for tracking reads and writes:
 * reads can be tracked by const accesses to the data pointer
 * writes can be tracked by mutable accesses to the data pointer
 * writes may also be tracked via autograd, using the same mechanism
   to bump version numbers

Note that we presently are only checking via autograd, since we don't
have const access to the data pointer, so we would be way too
aggressive if we assumed every access was a real write.

### Optimizations to the instrumentation
Technically, every tensor will require a shadow storage, because if we
were to create a view of a tensor and then create a lazy-copy, both
the original tensor and the view would have to share the shadow
storage, and thus it has to be created and shared before we ever even
know we needed it.

But we don't want to pay the memory cost of this since we don't expect
it to be that common. We can get around this by saving the original
shadow storage on the physical storage itself. We will have an
asymmetrical rule that states that any tensor that has a null shadow
storage will instead get its shadow storage from the physical
storage. This allows us to avoid refcount bumps on the shadow storage
as well as deferring any generation number bumps until we actually
have an outstanding copy on write.

The simulation instrumentation itself will be unnecessary to maintain
once we can transition to enabling lazy copies. This may happen after
the first that goes out which contains the instrumentation and the
user warning.

Future work
-----------
* enrich the warning by flagging reads/writes to data pointer after a
  big refactoring
* analyze violations of the warning and make a decision about whether
  we require any coordination about the BC change or if we should just
  let the warning run its course
* implement the actual copy on write
* simplify the compiler stack to no longer concern itself with this

Test Plan: Added new tests.

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/pytorch/pull/97175).
* #98536
* __->__ #97175

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

ezyang commented:
Needs more PR description

ezyang commented on c10/core/impl/copy_on_write_simulator.h:
```diff
@@ -0,0 +1,60 @@
+  // user's perspective, copies are fully distinct and require no
+  // external synchronization, so we need to ensure this field's
+  // concurrency is properly managed.
+  std::uint32_t storage_generation_;
```
Yeah, make this atomic.

dagitses replied:
True. But we could shave some memory off if we were able to use uint32_t for our refcounts in intrusive_ptr_target and more if we were to omit the weak refcount, which I don't believe we should need for this.

ezyang replied:
Fix your broken CI first :P

dagitses replied:
I'll promote to u64 in an amendment. Leaving this open until it's done.

dagitses replied:
done.

ezyang commented on c10/core/impl/copy_on_write_peer.cpp:
```diff
@@ -0,0 +1,43 @@
+
+/* static */ auto CopyOnWritePeer::get_generation(Storage const& storage)
+    -> std::uint32_t {
+  assert(storage);
```
don't use assert, we don't properly build with it. `TORCH_INTERNAL_ASSERT` for internal asserts that are cheap

dagitses replied:
Is there a reason we don't have a build with asserts enabled? I feel like I've been in a conversation about this multiple times at this point.

ezyang replied:
The point of `TORCH_INTERNAL_ASSERT` is that we can have it on all the time, which makes it substantially more useful for catching bugs in our binary distributions which aren't going to have asserts enabled.

dagitses replied:
all `assert` gone

ezyang commented on c10/core/impl/copy_on_write_peer.cpp:
```diff
@@ -0,0 +1,43 @@
+        "instead return a lazily copied tensor. If you wish to preserve the "
+        "aliasing properties, you should rewrite your reshape() as a view().");
+  }
+  tensor.copy_on_write_simulator_->bump_storage_generation();
```
you'll have to fix this to be atomic safe

dagitses replied:
should be done now.

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -2932,6 +2948,17 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   DispatchKeySet key_set_;
 
  private:
+  intrusive_ptr<impl::CopyOnWriteSimulator> copy_on_write_simulator_;
```
extra refcounted field, bloop.

dagitses replied:
Is there any way around that for the lifetime of this warning?

dagitses replied:
I think we can mitigate the impact of this by delaying population of this in the tensor until an actual copy-on-write is performed. Basically, we move this into the storage, and the storage will also track whether or not a copy-on-write has been performed. If we are in a situation where the storage reports that there's been a copy-on-write, but the tensor doesn't have a simulator, then it knows that it can inherit the "default" simulator since it is the simulator that all views would have had if we did this eagerly.

We can apply the same logic to defer any generation tracking as well.

dagitses replied:
If I am understanding this specific proposal correctly, this could also allow us to justify paying for a mutex to do this check with higher fidelity (synchronizing reads and writes to the generation numbers), since we ought to only need to use the mutex in situations where a copy-on-write would have actually been performed.

We can also throw all this checking away once we've warned once, since we'll never warn again.

So the state diagram I imagine is this:
1) tensor is created. views of this tensor may also be created. There's a single storage, and it says that there's no copy-on-writes going on. Each tensor has a null copy-on-write simulator.
2) reshape would have created a CoW tensor. The new "copy" has the aliased storage, but it also has its own CoW simulator inside the tensor. We create a default CoW simulator inside the storage: this default value is only used by all of the views that pre-existed the copy-on-write.
2b) all read/write operations at this point to any of the view families that share this storage are now going to do all the fun generation tracking.
3) if at any point we trigger the warning, then we can discard all the state for this storage. If we want to go further, we can set a global telling us to never create any copy-on-write state again, assuming we are doing a TORCH_WARN_ONCE.

Note that we may want to distinguish reads after writes to a different copy and writes after writes to a different copy (i.e. two different warnings).

dagitses replied:
OK, unfortunately we're a little late for that. I have implemented this but it does still require a mutex acquisition even when we aren't using this. That can probably be downgraded to an atomic read eventually.

Testing this new configuration in CI, here we go!

dagitses replied:
OK, seems like we're ok with the approach here. We can drop this to not be refcounted as described in the conversation about templating on intrusiveness.

ezyang commented:
unsubbing for now, shout when more review

dagitses commented:
@ezyang This ought to be ok for review now. There's one outstanding failure, also with compiling c++ extensions in windows not using C++17: https://github.com/pytorch/pytorch/actions/runs/4500159435.

I'll fix that one as well.

I am severely concerned about the disabled assert in `cow::State::maybe_bump()`. I suspect that either I'm missing some constructor called by a view or that some Tensor is having it's storage surgically altered. I do not want to submit it as-is, but I suspect whatever remedy is required will be rather limited.

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -723,12 +723,15 @@ def spawn(cmd):
                             cflags = ['-Xcompiler', flag] + cflags
                         for ignore_warning in MSVC_IGNORE_CUDAFE_WARNINGS:
                             cflags = ['-Xcudafe', '--diag_suppress=' + ignore_warning] + cflags
+                        append_std17_if_no_std_present(cflags)
```
Move this to a separate C++17

dagitses replied:
This is, sapling just stacks differently from ghstack, it only hides the parent commit in ReviewStack. Sorry :(

ezyang commented on c10/core/impl/cow/simulator.h:
```diff
@@ -0,0 +1,60 @@
+//
+// We use a monotonically increasing generation number to track
+// modifications to storage.
+template <bool intrusive = true>
```
Why do we template on intrusive?

dagitses replied:
I was seeing ASan issues with stack allocated intrusive_ptr_target subclass. That is expected?

We don't need the default shadow storage in the cow::StateMachine to be refcounted, since its lifetime is bound to that instance.

Come to think about it, do we need to refcount any of them? We could just make them live as long as the storage lives, and just allocate them in a std::deque or set-like object that doesn't invalidate. It would be probably slight tradeoff of peak memory vs refcount bumps.

An upside of doing that is that we could enhance the assert in maybe bump to verify the association of each shadow storage with the physical storage.

ezyang replied:
This probably doesn't matter. Our cost model is that reshape() is expensive, but everything else is cheap now.

ezyang commented on c10/core/impl/cow/simulator.h:
```diff
@@ -0,0 +1,60 @@
+// tensor would have been created and whether or not a logical copy
+// uses the storage after a different copy has modified it. So in
+// theory, it can identify reads/writes to copies after a write to a
+// different copy that is presently an alias.
```
nit: want to know what a simulator semantically represents. In this case, it would be the "shadow storage" representing what the storage would look like after the CoW change

dagitses replied:
reworded this, hopefully more clear.

ezyang commented on c10/core/impl/cow/README.md:
```diff
@@ -0,0 +1,65 @@
+implementation to *always* copy, but copy lazily upon modification if
+we can represent the output as a view.
+
+Backwards imcompatibiility
```
incompatibility

dagitses replied:
names are updated

ezyang commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+
+// Responsible for managing the copy-on-write simulation state
+// machine.
+class C10_API State {
```
nit: prefer a more descriptive name here

dagitses replied:
updated to cow::StateMachine.

ezyang commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+namespace c10::impl::cow {
+
+// Responsible for managing the copy-on-write simulation state
+// machine.
```
An abstract description of the state machine here would be helpful

ezyang replied:
apparently the state machine where you stop tracking after a warning has been raised. But not clear if this is worth implementing?

dagitses replied:
Agreed. That will come in the next iteration. I'm going to formally define an enum with the states and transitions and document each individually.

The null state is important. I want to drop the cost of the null state to an atomic read. In the common case, we can make it an uncontested read that will indicate that there's no lazy copies to be concerned about. This will be much better than the write to acquire a mutex.

dagitses replied:
OK, the state machine has manifested itself.

ezyang commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+  // simulate_copy_on_write on this instance.
+  auto maybe_bump(cow::Simulator* maybe_simulator) -> void;
+
+  // Simulates a copy on write.
```
`simulate_lazy_copy`

dagitses replied:
switched to lazy copy

ezyang commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+
+  // Simulates a copy on write.
+  //
+  // maybe_simulator comes from the tensor that would be receiving a
```
nit: s/receiving/was the source of the lazy copy/

dagitses replied:
reworded

ezyang commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+  // maybe_simulator comes from the tensor that would be receiving a
+  // copy on write. If non-null, then the result gets the generation
+  // number from this value.
+  auto simulate_copy_on_write(cow::Simulator* maybe_simulator)
```
Worth saying why this is not necessarily the same as default simulator

dagitses replied:
reworded

ezyang commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+  //
+  // maybe_simulator comes from the tensor that would be receiving a
+  // copy on write. If non-null, then the result gets the generation
+  // number from this value.
```
The lifecycle of default_simulator from null to non-null should be clearly described, it's not easy to figure out what this is doing here

ezyang commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+  // Gets the current storage generation.
+  auto storage_generation() -> std::uint64_t;
+
+  // Bumps the generation if simulator is non-null. If non-null,
```
When do you bump the generation? When you mutate any tensor that aliases with this storage

ezyang commented on c10/core/impl/cow/state.cpp:
```diff
@@ -0,0 +1,59 @@
+    // TODO: This check should probably be here, but it is actually
+    // triggering in the StaticRuntime.autogen_inner test.
+    //
+    // TORCH_INTERNAL_ASSERT(maybe_simulator == nullptr);
```
It's being violated because you haven't handled what happens if you do `set_`.

ezyang replied:
set_ that takes Storage is not possible to do precisely. set_ with Tensor can be done precisely.

ezyang replied:
Resolution: wipe this to nullptr

dagitses replied:
done for set with storage, not set with tensor yet. Set with tensor may be precise because we can just copy over the value of shadow_storage_ from the source tensor?

ezyang commented:
I need some real world data on the PR. For example, how often does the warning trigger on the test suite? How about torchbenchmark? How can we make sure it's actually OK to deploy the BC-breaking change in the end?

ezyang commented on aten/src/ATen/core/copy_on_write.cpp:
```diff
@@ -0,0 +1,19 @@
+  }
+
+  c10::TensorImpl* tensor_impl = tensor.unsafeGetTensorImpl();
+  if (tensor_impl == nullptr) {
```
This feels like it should be an assert failure instead

dagitses replied:
Is this even possible? Given that TensorBase's member uses UndefinedTensorImpl as its NullType?

dagitses replied:
done

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1570,10 +1576,6 @@ Tensor reshape_symint(const Tensor& self, c10::SymIntArrayRef proposed_shape) {
   }
 
-  if (self.is_contiguous() && !self.is_mkldnn()) {
-    return self.view_symint(proposed_shape);
```
This may break some unbacked SymInt tests, although after I added this branch we changed the strategy so maybe this is dead. Safer to do this separately.

ezyang commented on test/test_copy_on_write.py:
```diff
@@ -0,0 +1,52 @@
+        t.add_(torch.arange(4))
+        # Even writing twice to t is fine.
+        t.add_(torch.ones(4))
+
```
Is it OK to take out a view from u without "reading" it?

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -184,6 +177,44 @@ TensorImpl::TensorImpl(
   // Caffe2 operators create Storages with default devices.
 }
 
+TensorImpl::TensorImpl(
```
TODO: need to audit these constructors more carefully in code review

ezyang commented:
Algorithmically this looks really promising! One word on TensorImpl is also acceptable, and the perf optimization (I know I wasn't originally a fan of putting it in) seems very good.

The higher order bit for next steps on this PR is to validate how many warnings we get on representative codebases, to make sure it's not tons of warnings.

dagitses commented on c10/core/impl/cow/state.h:
```diff
@@ -0,0 +1,49 @@
+class C10_API State {
+ public:
+  // Gets the current storage generation.
+  auto storage_generation() -> std::uint64_t;
```
done

dagitses commented on c10/core/impl/cow/simulator.cpp:
```diff
@@ -0,0 +1,33 @@
+  std::lock_guard<std::mutex> lock(mtx_);
+  TORCH_INTERNAL_ASSERT(
+      storage_generation_ != std::numeric_limits<std::uint64_t>::max());
+  return ++storage_generation_;
```
Done.

It should have always been atomic. Early in the design I was being cautious about ensuring that we always write and read both generation numbers atomically. I turned out to be overly cautious, since we update them both under the physical storage's mutex and we always have that mutex held whenever we need to read them atomically.

dagitses commented:
OK this is worth looking at again. There's going to be an enhancement to the state machine to make the states and transitions more formal and also to reduce the cost in the null state to an uncontested atomic read.

My review style is to leave it up to the reviewer to resolve any issues, since it ought to be up to them to decide if they are satisfied with the outcome. So please resolve any issues you don't wish to discuss further.

ezyang commented:
I'm waiting on real world experiments before doing more review. The code is basically fine, but I want to know if we can actually land this.

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -2932,6 +2961,27 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  //
+  // This is asymmetrical, but it allows us to avoid the allocation
+  // and any refcount bumps until we actually need them.
+  intrusive_ptr<impl::cow::ShadowStorage> shadow_storage_;
```
It would be good to have a direct pointer to the relevant README, so that someone who shows up here looking to axe out fields can figure out where to go

dagitses replied:
added

ezyang commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,104 @@
+namespace c10::impl::cow {
+
+// Responsible for managing the copy-on-write simulation state
+// machine.
```
but what is the cow simulation state machine 🤔

dagitses replied:
hopefully the expanded comment covers this

ezyang commented on c10/core/impl/cow/shadow_storage.h:
```diff
@@ -0,0 +1,70 @@
+// Note that we templatize on whether or not we are eligible to be
+// allocated into an intrusive_ptr. It *seems* as though ASan is
+// unhappy if we stack-allocate an object that inherits from
+// c10::intrusive_ptr_target.
```
What exactly is the ASAN error?  See

```
// Note [Stack allocated intrusive_ptr_target safety]
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// A well known problem with std::enable_shared_from_this is that it
// allows you to create a std::shared_ptr from a stack allocated object,
// which is totally bogus because the object will die once you return
// from the stack.  In intrusive_ptr, we can detect that this has occurred,
// because we set the refcount/weakcount of objects which inherit from
// intrusive_ptr_target to zero, *unless* we can prove that the object
// was dynamically allocated (e.g., via make_intrusive).
//
// Thus, whenever you transmute a T* into a intrusive_ptr<T>, we check
// and make sure that the refcount isn't zero (or, a more subtle
// test for weak_intrusive_ptr<T>, for which the refcount may validly
// be zero, but the weak refcount better not be zero), because that
// tells us if the object was allocated by us.  If it wasn't, no
// intrusive_ptr for you!
```

dagitses replied:
i removed the templating on intrusiveness so we'll see if it is still happening.

dagitses replied:
This did not recur on the latest manifestation. I'm going to assume that I had some other programming error that caused this that no longer exists.

ezyang commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,104 @@
+  // because of torchdeploy incompatibility.
+  //
+  // See https://github.com/pytorch/multipy/issues/312
+  c10::optional<cow::ShadowStorageNonIntrusive> default_shadow_storage_;
```
This is a lot of fields. Mutex is like 80 bytes, and then you also have four more words worth of stuff. Making matters worse, this struct is included inline in StorageImpl, which means we're paying this cost on every storage allocation. This will definitely affect Caffe2 perf.

ezyang replied:
We talked over VC about whether or not it was actually necessary mutex protect these fields. My intuition is that it is not necessary. Additionally, use of c10::optional is costing you another word on default_shadow_storage_, but in fact what's the problem with just always having a ShadowStorageNonIntrusive and using state_id to control whether or not it is set or not?

dagitses replied:
I added the ability to disable this on the TensorImpl side with the `PYTORCH_INSTRUMENT_COW_TENSOR` preprocessor definition. I implemented this to take up no storage on TensorImpl if it's not defined. Presumably we want the same here?

I tested this earlier, std::mutex is 40 bytes in both GCC and Clang.

We can cut down on the size a little. We can make the state_id_ one byte (assuming atomic supports that). We can then also make our generations 32-bit. And we can drop the `c10::optional` around the `default_shadow_storage_` and make its presence be contingent on the value of `state_id_`.

All together we should be at:
40 bytes: mutex
4 bytes: generation
4 bytes: default shadow_storage
1 byte: state id

The alignment of mutex is 8, so we don't gain anything by decreasing state id's size.

Is there a lighter weight mutex we can use?

Alternatively, I assume you are suggesting that we heap allocate the state machine in StorageImpl on demand, so we only pay the weight when there is a lazy copy active. We can easily do that.

dagitses replied:
OK, this is in place. StorageImpl only pays a pointer size when this is not in use. I didn't implement the empty-base class optimization for it yet, but I can add that too, it's really easy.

dagitses replied:
OK, so let's discuss whether or not we still need a Mutex here. Let's assume it doesn't exist.

In bump, we atomically increase the physical generation then, in a separate atomic operation, we atomically update the shadow storage of the tensor being written to.

Tensors that share a shadow storage are tensors that are views of one another, so they may not be mutated concurrently. However, tensors who share a "real" storage are logically lazy copies, and they may be mutated concurrently.

So if a concurrent write is taking place to two lazy copies, in the future world, at worst, both will materialize their own copies, and there's no issue.

In the legacy world, these are both physically views and they will conflict. We would like a warning to take place. The atomic bumps to the physical generation will not race, so one of them will see N + 1 and the other will see N + 2.

We should be able to safely just atomically set the shadow generations for each and one of them will warn.



Now let's see what will happen with a read that is racing with a write. Again, this is forbidden if these share a shadow storage, because a user is required to synchronize writes and reads to logically aliasing tensors.

So let's have a write to one lazy copy and a read from another racing.

Read thread:
R1) atomically read physical generation (shared with write)
R2) atomically read shadow generation (distinct from write)

Write thread:
W1) atomically increment physical generation (shared with read)
W2) atomically update shadow generation (distinct from read)

The order of R2 and W2 don't matter, so let's ignore those.

W1 -> R1: The reader sees the updated physical generation and will warn correctly.
R1 -> W1: The reader sees the old physical generation and will not warn. This may or may not be a false negative, depending on how the actual write and reads are sequenced. This is, however, sketchy in the old regime.

Nevertheless, the important thing is that there are no false positives. So I think we can safely drop the mutex. What do you think?

dagitses replied:
#98504 for an idea of what i mean

ezyang commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,104 @@
+  // Guards all the state.
+  std::mutex mtx_;
+  // How many writes have been applied to the storage.
+  std::uint64_t physical_generation_;
```
Is it possible for physical generation to be nonzero and state id to be inactive?

ezyang commented on c10/core/impl/cow/README.md:
```diff
@@ -0,0 +1,147 @@
+storage will instead get its shadow storage from the physical
+storage. This allows us to avoid refcount bumps on the shadow storage
+as well as deferring any generation number bumps until we actually
+have an outstanding copy on write.
```
It's worth talking about when this instrumentation *can be deleted.*

dagitses replied:
added

dagitses commented on c10/core/TensorImpl.cpp:
```diff
@@ -184,6 +177,44 @@ TensorImpl::TensorImpl(
+  }
+}
+
+intrusive_ptr<TensorImpl> TensorImpl::take_view() const {
```
i haven't done this yet. do you recommend expanding my unit test with all the view ops to make sure they are covered?

ezyang replied:
We should have OpInfos for view ops, you just have to make sure you send them CoW'd tensors in a custom test.

dagitses replied:
i've migrated the calls of ImplType::VIEW to use take_view. Right now i'm not supporting tensor subclasses, so those calls still exist.

ezyang replied:
As I mentioned in another comment, I think you need to handle tensor subclasses. And because you need to handle tensor subclasses, I think the `take_view` method as done here is not good design. You want to set everything up in the constructors. What would have been more logical to me would have been to modify the VIEW constructor to accept shadow storage, and then update all the call sites to do the right thing.

ezyang commented:
Can you avoid force pushing to this PR? It makes it difficult to see what the incremental changes over time are.

ezyang commented on c10/core/Storage.h:
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <c10/core/StorageImpl.h>
+#include <c10/core/impl/cow/shadow_storage.h>
```
Shouldn't be necessary, you should IWYU?

dagitses commented:
> 

sorry that's what sapling does. If I switch to internal first and ghexport, will that work better?

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+cow::StateMachine::StateMachine() : state_rep_(StateId::initial) {}
+
+cow::StateMachine::~StateMachine() {
+  // If we created an impl, we are responsible for clenaing this up
```
nit: cleaning

dagitses replied:
done

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+  impl->maybe_bump(maybe_shadow_storage);
+}
+
+auto cow::StateMachine::simulate_lazy_copy(
```
allocates impl

ezyang commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,176 @@
+  };
+
+  // The current representation of the storage instance.
+  std::atomic<Rep> state_rep_;
```
OK, this is good, thank you for adjusting this.

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+// This is a less sketchy version of reinterpret_cast.
+//
+// See https://en.cppreference.com/w/cpp/numeric/bit_cast for more
+// information as well as the source of our implementations.
```
Put it in c10/utils?

dagitses replied:
this 3 diff stack addresses that: D44712290

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+  // Guards all the state.
+  std::mutex mtx_;
+  // How many writes have been applied to the storage.
+  cow::StateMachine::Generation physical_generation_ = 0;
```
There is also a generation on the ShadowStorageImpl as well, so it would be helpful to point out what exactly the distinction between this and the generation stored on the default shadow storage is

dagitses replied:
Please take a another look.

ezyang commented on c10/core/impl/cow/README.md:
```diff
@@ -0,0 +1,152 @@
+storage will instead get its shadow storage from the physical
+storage. This allows us to avoid refcount bumps on the shadow storage
+as well as deferring any generation number bumps until we actually
+have an outstanding copy on write.
```
This seems slightly out of date now

dagitses replied:
I re-read the whole thing, I don't see what is out of date.

ezyang commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,176 @@
+    // alignment bits of a pointer to hold this state id. In the case
+    // where we have an actual impl, the state id will be active,
+    // hence the alignment bits will be 0.
+    active = 0,
```
nit: you probably want some sort of assert that says ensures this enum doesn't get too big, otherwise you'll run out of bits.

dagitses replied:
moot!

ezyang commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,176 @@
+  // will call this on the Tensor receiving the mutation, and it will
+  // propagate through its storage to here. That tensor will send its
+  // shadow storage here, but it will be null if it was the original
+  // tensor or a view derived from it.
```
I think I asked about this before, but this comment still doesn't make it clear why you also have to "send its shadow storage here". Maybe something like "Note that this doesn't bump the *default* shadow storage impl; instead it bumps the generation on the real storage and the generation on the passed in shadow storage (which may or may not be the default shadow storage impl)."

ezyang commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,176 @@
+  // Responsible for cleaning up any implementation details.
+  ~StateMachine();
+
+  // Gets the current generation of the physical storage.
```
nit: Worth saying that no tracking happens UNTIL we simulate a lazy copy, so you're not allowed to use this to track mutations to the storage otherwise

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+    // TODO: This check should probably be here, but it is actually
+    // triggering in the StaticRuntime.autogen_inner test.
+    //
+    // TORCH_INTERNAL_ASSERT(maybe_shadow_storage == nullptr);
```
So what happened to the comment here?

ezyang replied:
https://github.com/pytorch/pytorch/pull/97175#discussion_r1147631124

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+    TORCH_INTERNAL_ASSERT(
+        new_rep.as_state_id() == cow::StateMachine::StateId::active);
+    // We won the race to initializing: we will be the only thread
+    // that may transition to active.
```
Instead of spinning on initializing, can't you just speculatively allocate StateMachine::Impl and then throw it out if you lose the write race?

ezyang replied:
This reduces the number of CEX from two to one, and also avoids the need for spinlocking.

dagitses replied:
Good idea and nice subsequent simplification (in progress) that follows.

dagitses replied:
done

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+  Generation physical_generation = ++physical_generation_;
+  Generation shadow_generation = apply_to_active_shadow_storage(
+      maybe_shadow_storage, default_shadow_storage_, [](auto& shadow_storage) {
+        return shadow_storage.bump_generation();
```
It's a little weird to do the shadow storage as a bump, rather than assigning physical generation to it (and doing the warn check before you do the assignment.

The difference here is what happens if you do:

```
y = x.reshape(...)
x.add_(2)
y.add_(2)
```

The way you have this implemented, after the two mutations, both x and y's shadow storage are "one generation short" of physical generation, so everyone is going to complain on subsequent reads! But it seems more humane to warn on the y.add_(2) (because writes count as reads) and then just assume that y is "up to date" and any reads on it are OK.

dagitses replied:
nice. done.

ezyang commented on c10/core/impl/cow/state_machine.cpp:
```diff
@@ -0,0 +1,240 @@
+    -> StateMachine::Impl* {
+  bool seen_initializing = false;
+  do {
+    Rep state_rep = state_rep_;
```
I'd appreciate an explicit load() here

dagitses replied:
fair enough. done.

ezyang replied:
If you take my suggestion about going to only one CEX, `state_rep_` now monotonically goes from nullptr to a pointer (in fact, you don't need Rep anymore, because you no longer have initializing state, and you can use nullptr as the uninitailized sigil). Because the transition is monotonic, you don't need a seq cst load. In fact, for some uses of `get_impl`, you don't need an atomic read at all, because you don't care if you race with another thread establishing the default storage impl or not. This is specifically in the mutation detection case: you are always happy to report that there is no problem, because if you're racing with a writing thread, well, you're racing with a writing thread.

ezyang replied:
And for other uses of read, a more relaxed read is OK. I think consume is good enough.

dagitses replied:
good point for the read before exchange. For the read in maybe_bump(), I agree that consume should do the trick since we just need to ensure that we don't see a null after a simulate is created. If they raced in a real copy-on-write world, then that means a tensor would have been having a read or write concurrent with a write to the same tensor or one its views. That's already not supported so I we don't have to be any stricter than that.

the CEX can be memory_order_acq_rel yes?

done.

ezyang commented:
The main blocker for my approve is that not all views are propagating shadow storages correctly at the moment.

dagitses commented on c10/core/impl/cow/state_machine.h:
```diff
@@ -0,0 +1,176 @@
+    // Subtle! This has the value of 0 because we are using the
+    // alignment bits of a pointer to hold this state id. In the case
+    // where we have an actual impl, the state id will be active,
+    // hence the alignment bits will be 0.
```
moot!

dagitses commented:
There is a segfault in XLA. I'm not worrying about it too much until I see what the behavior is in #98504.

https://github.com/pytorch/pytorch/actions/runs/4629524354/jobs/8190667428

dagitses commented on c10/core/impl/cow/README.md:
```diff
@@ -0,0 +1,152 @@
+
+But we don't want to pay the memory cost of this since we don't expect
+it to be that common. We can get around this by saving the original
+shadow storage on the phsyical storage itself. We will have an
```
fixed thanks

dagitses commented:
> The main blocker for my approve is that not all views are propagating shadow storages correctly at the moment.

This may no longer be true. I updated that and added an opinfo aspect that transforms inputs into lazy copies.

Commit:
warn on future reshape alias mutation violations

Summary:
Copy-on-Write Storage
=====================

Motivation
----------
PyTorch inherited from NumPy an optimization on the reshape function
that produces a view as an output if it can be represented as
such. This complicates the work of the PyTorch compilation stack
because it needs to understand the representation of the input in
order to understand if the output will be a copy or an alias.

The compilation stack would rather not be concerned with such details,
motivating the Stride Agnostic PyTorch project.

To address reshape specifically, we wish to simplify the
implementation to *always* copy, but copy lazily upon modification if
we can represent the output as a view.

Implementation plan
-------------------
We have not implemented copy-on-write tensors yet, because this is a
backward incompatible change (see Backward Incompatible section
below). But this is the design.

A copy-on-write tensor, also known as a lazily-copied tensor, is
initially going to be created by an operation like reshape which would
historically have created a view. We wish to maintain the performance
of the view but drop the aliasing aspect. Note that there is desire
for copy-on-write tensors outside of reshape, so we will likely also
want to add a public operator that can do this. It could be named
"lazy_clone" or something similar.

The core tenet of the design is that we wish to maintain the invariant
that Tensors alias if and only if they share a storage. Thus when we
create a lazy-copy, we will need to create a new storage. We also will
have to modify the source storage, since it also must now be lazily
copied.

The algorithm for creating a new copy-on-write tensor (and also
converting the source to copy-on-write) is as follows:

```
def lazy_clone(src: Tensor) -> Tensor:
    # First ensure that the source has a copy-on-write enabled storage.
    # We implement this using a custom context on the DataPtr. The
    # source tensor might already have this enabled, but if it doesn't,
    # convert it.
    if not has_copy_on_write_context(src.storage().storage_impl()):
        if not(wrap_with_copy_on_write_context(src.storage().storage_impl())):
            # For whatever reason, we weren't able to wrap the DataPtr.
            # We have to do an eager copy.
            return src.clone()

    new_storage = fork_copy_on_write_storage(src.storage()) # can't fail
    # Now just create a new tensor using the new storage.
    return new Tensor(storage=new_storage, sizes_and_strides_like=src)
```

That's the high level algorithm. The copy-on-write context that we
introduce is morally just a refcount on the underlying physical data
pointer. Each unique storage represents a set of tensors that share a
view and thus will hold a single refcount on the context.

Now we just need to intercept writes to the storage and materialize
them. We can use a few mechanisms to do this:

1) autograd knows which operators write to which tensor inputs. We can
   materialize at that point when autograd is enabled.
2) if autograd is not enabled, we can introduce a new dispatch key
   that does the same trick
3) we can also materialize whenever there's mutable access to the data
   pointer through any of `at::Tensor`, `c10::TensorImpl`,
   `c10::Storage`, `c10::StorageImpl`, or `c10::DataPtr`. With the
   current codebase, this will be too aggressive, but we will refactor
   to have a minimal set of mutable accesses.

Backwards incompatibiility
--------------------------
Changing reshape to produce a copy is a backwards incompatible change,
because users could be relying on the aliasing behavior, intentionally
or not.

For one release, rather than returning copy-on-write tensors, we
instead warn when users have triggered behavior in their program that
relies on the aliasing of the output and input.

To do this, we must simulate the behavior in a backward compatible
way. To remain backward compatible, the aliases must preserve the
invariant that they have the same storage. This is a big deviation
from the design detailed above. To get closer to the real design and
implementation, we introduce a new `c10::TensorImpl` level concept
called "Shadow Storage". The shadow storage represents what the
storage would have looked like the view actually been a lazy copy.

In the instrumented world we thus maintain the formal invariant that
tensors that alias share a storage. But we have a new invariant:
tensors that are lazy-copies of each other will share a shadow
storage.

So what do we warn on? We warn if there is a write to a tensor in a
set that shares a shadow storage followed by a read or a write to a
different tensor that shares a physical storage but has a different
shadow storage. In the real implementation, the first write would have
triggered a copy, forever cleaving the two sets of tensors, but in the
current world we instead had behavior that relied on the view-ness of
the output of reshape.

We can track these violations simply by adding a generation number to
the shadow and physical storages, updating them both on writes and
observing if a read or write ever encounters values that are out of
sync.

We have a few mechanisms for tracking reads and writes:
 * reads can be tracked by const accesses to the data pointer
 * writes can be tracked by mutable accesses to the data pointer
 * writes may also be tracked via autograd, using the same mechanism
   to bump version numbers

Note that we presently are only checking via autograd, since we don't
have const access to the data pointer, so we would be way too
aggressive if we assumed every access was a real write.

### Optimizations to the instrumentation
Technically, every tensor will require a shadow storage, because if we
were to create a view of a tensor and then create a lazy-copy, both
the original tensor and the view would have to share the shadow
storage, and thus it has to be created and shared before we ever even
know we needed it.

But we don't want to pay the memory cost of this since we don't expect
it to be that common. We can get around this by saving the original
shadow storage on the physical storage itself. We will have an
asymmetrical rule that states that any tensor that has a null shadow
storage will instead get its shadow storage from the physical
storage. This allows us to avoid refcount bumps on the shadow storage
as well as deferring any generation number bumps until we actually
have an outstanding copy on write.

The simulation instrumentation itself will be unnecessary to maintain
once we can transition to enabling lazy copies. This may happen after
the first that goes out which contains the instrumentation and the
user warning.

Future work
-----------
* enrich the warning by flagging reads/writes to data pointer after a
  big refactoring
* analyze violations of the warning and make a decision about whether
  we require any coordination about the BC change or if we should just
  let the warning run its course
* implement the actual copy on write
* simplify the compiler stack to no longer concern itself with this

Test Plan: Added new tests.

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -573,7 +587,24 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
       const caffe2::TypeMeta data_type,
       c10::optional<c10::Device>);
 
+  // Private constructor for lazily copying/viewing tensors.
```
I think you should tag this constructor with the VIEW tag. In particular, it's non-obvious that this constructor DOESN'T copy sizes/strides. (This is the desired behavior, but one could easily be deceived by the fact that this takes a Tensor)

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -512,7 +518,15 @@ class C10_TensorImpl_Size_Check_Dummy_Class;
+      //
+      // See c10/core/impl/cow/README.md for a detailed description of
+      // what shadow storage is.
+      public impl::cow::ShadowStorageMixin {
```
Does this really want to be public inheritance?

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -104,24 +104,16 @@ TensorImpl::TensorImpl(
+          std::move(storage),
+          key_set,
+          data_type,
+          /*shadow_storage=*/nullptr) {}
```
I think you need to do the subclasses too. Once you've expunged all the old VIEW sites you should delete this constructor entirely, as it is a footgun that makes it easy to incorrectly propagate shadow storage.

ezyang commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10093,6 +10093,7 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                DecorateInfo(unittest.expectedFailure, 'TestBwdGradients', 'test_fn_grad'),
                # Jacobian mismatch
                DecorateInfo(unittest.expectedFailure, 'TestBwdGradients', 'test_fn_gradgrad'),
+               DecorateInfo(unittest.skip('output is non-deterministic'), 'TestCommon', 'test_lazy_copy_samples'),
```
How come?

ezyang commented on test/test_ops.py:
```diff
@@ -652,6 +652,97 @@ def test_noncontiguous_samples(self, device, dtype, op):
+                else tuple(n_inp) + n_args
+            )
+
+            # Filter the elemnts that are tensors that require grad
```
nit: element

ezyang commented on test/test_ops.py:
```diff
@@ -652,6 +652,97 @@ def test_noncontiguous_samples(self, device, dtype, op):
+    @onlyNativeDeviceTypes
+    @suppress_warnings
+    @ops(op_db, allowed_dtypes=(torch.float32, torch.long, torch.complex64))
+    def test_lazy_copy_samples(self, device, dtype, op):
```
I'm not... sure what you are intending to test here? The big ask is to check that view operations preserve non-standard shadow storage. This test is testing that every operator gives the same result if the storage has shadow storage vs not but... like, why would we expect this not to hold? We didn't change any of the Storage logic itself.

ezyang replied:
In other words, this test is fairly expensive, seems to expose flakiness (e.g., cov), and doesn't even seem to test what is likely to be wrong. It doesn't seem like good ROI.

kurtamohler commented:
I'm closing this because it will be superseded by #126129
